addons/mail/static/src/messaging/entity/composer/composer.js

```diff
@@ -220,6 +255,76 @@ function ComposerFactory({ Entity }) {
             return [...new Set(allPartnerIds)];
         }
 
+        _getFirstMentionSuggestion() {
+            this.active_id = this.mentionSuggestions[0].id;
+        }
+
+        _getLastMentionSuggestion() {
+            this.active_id = this.mentionSuggestions[this.mentionSuggestions.length - 1].id;
+        }
+
+        async _getMentionSuggestions(mentionWord) {
+            this.active_id = 0;
+            const mentions = await this.env.rpc({
+                model: 'res.partner',
+                method: 'get_mention_suggestions',
+                kwargs:
+                    {
+                        limit: 3,
+                        search: mentionWord,
+                    },
+            });
+
+            this.unlink({ mentionSuggestions: null });
+            for (const mention of mentions[0]) {
+                const { email, id, name } = mention;
+                const mentionSuggestion = this.env.entities.Partner.insert({ email, id, name });
+                this.link( {mentionSuggestions: mentionSuggestion} )
+            }
+
+            if (this.mentionSuggestions.length > 0) {
+                this.active_id = this.mentionSuggestions[0].id;
+            }
+        }
+
+        _getNextMentionSuggestion() {
+            const active_index = this.mentionSuggestions.findIndex(mentionSuggestion => mentionSuggestion.id == this.active_id);
+            if (active_index !== this.mentionSuggestions.length - 1) {
+                this.active_id = this.mentionSuggestions[active_index + 1].id;
+            } else {
+                this.active_id = this.mentionSuggestions[0].id;
+            }
```

Comment:
index of items in a relation is unreliable. That's because relation is a list for `x2many`, but order is arbitrary by design. If you need order, you should make it yourself, i.e.g with an attribute that acts as a list of ordered local ids, and/or a getter that returns ordered items based on their properties like id or sequence, etc...